snuba/subscriptions/verifier.py

```diff
@@ -99,3 +113,200 @@ def close(self, timeout: Optional[float] = None) -> None:
+                >= self.__timestamp_low_watermark + self.__threshold_sec
+            ):
+                self.__metrics.increment("stale_message")
+            return
+
+        # Increment counters
+        store = self.__stores[topic]
+        store[item.timestamp] += 1
+
+        # Record metrics and advance the low watermark
+        low_watermark = self.__timestamp_high_watermark - self.__threshold_sec
+
+        if low_watermark > self.__timestamp_low_watermark:
+            self._record(low_watermark)
+            self.__timestamp_low_watermark = low_watermark
+
+    def _record(self, until: int) -> None:
+        """
+        Records metrics up to and including `until` then drop recorded elements.
+        """
+        while True:
+            next_orig_ts = next(iter(self.__stores[ResultTopic.ORIGINAL]), None)
+            next_new_ts = next(iter(self.__stores[ResultTopic.NEW]), None)
+
+            if next_orig_ts is not None and next_orig_ts > until:
+                next_orig_ts = None
+            if next_new_ts is not None and next_new_ts > until:
+                next_new_ts = None
+
+            # Set next_ts
+            if next_orig_ts is None and next_new_ts is None:
+                next_ts = None
+            elif next_orig_ts is None:
+                next_ts = next_new_ts
+            elif next_new_ts is None:
+                next_ts = next_orig_ts
+            else:
+                next_ts = min([next_orig_ts, next_new_ts])
+
+            # Nothing else to remove
+            if next_ts is None:
+                return
+
+            orig_count = self.__stores[ResultTopic.ORIGINAL].pop(next_ts, 0)
+            new_count = self.__stores[ResultTopic.NEW].pop(next_ts, 0)
+            self.__metrics.increment("result_count_diff", abs(new_count - orig_count))
+
+
+class CountResults(ProcessingStrategy[SubscriptionResultData]):
+    """
+    Records a metric with the result counts on the new vs original
+    result topics.
+    """
+
+    def __init__(
+        self,
+        orig_result_topic: Topic,
+        new_result_topic: Topic,
+        threshold_sec: int,
+        metrics: MetricsBackend,
+        commit: Callable[[Mapping[Partition, Position]], None],
+    ):
+        self.__orig_result_topic = orig_result_topic
+        self.__new_result_topic = new_result_topic
+        self.__commit = commit
+
+        self.__store = ResultStore(threshold_sec, metrics)
+
+    def poll(self) -> None:
+        pass
+
+    def submit(self, message: Message[SubscriptionResultData]) -> None:
+        assert not self.__closed
+
+        message_topic = message.partition.topic
+        store_key = {
+            self.__orig_result_topic: ResultTopic.ORIGINAL,
+            self.__new_result_topic: ResultTopic.NEW,
+        }[message_topic]
+
+        self.__store.increment(store_key, message.payload)
+
+        # Immediately commit offsets
+        self.__commit({message.partition: Position(message.offset, message.timestamp)})
```

Comment:
Are you not committing only one offset here ?
Also you would be committing around 100 times per second per topic as you commit for every query executed. That's too much. Please batch the commits.

Reply:
Updated to throttle to 1 a second.